hnn_core/gui/_viz_manager.py

```diff
@@ -0,0 +1,598 @@
+"""HNN GUI visualization management tool"""
+
+# Authors: Huzi Cheng <hzcheng15@icloud.com>
+
+import copy
+import io
+from functools import partial
+
+import matplotlib
+import matplotlib.pyplot as plt
+import numpy as np
+from IPython.display import display
+from ipywidgets import (Box, Button, Dropdown, FloatText, HBox, Label, Layout,
+                        Output, Tab, VBox, link)
+
+from hnn_core.gui._logging import logger
+from hnn_core.viz import plot_dipole
+
+
+_fig_placeholder = 'Run simulation to add figures here.'
+
+_plot_types = [
+    'current dipole',
+    'layer2 dipole',
+    'layer5 dipole',
+    'aggregate dipole',
+    'input histogram',
+    'spikes',
+    'PSD',
+    'spectogram',
+    'network',
+]
+
+_spectrogram_color_maps = [
+    "viridis",
+    "plasma",
+    "inferno",
+    "magma",
+    "cividis",
+]
+
+
+fig_templates = {
+    "2row x 1col (1:3)": {
+        "kwargs":
+        "gridspec_kw={\"height_ratios\":[1,3]}",
+        "mosaic": "00\n11",
+    },
+    "2row x 1col (1:1)": {
+        "kwargs":
+        "gridspec_kw={\"height_ratios\":[1,1]}",
+        "mosaic": "00\n11",
+    },
+    "1row x 2col (1:1)": {
+        "kwargs":
+        "gridspec_kw={\"height_ratios\":[1,1]}",
+        "mosaic": "01\n01",
+    },
+    "single figure": {
+        "kwargs": "",
+        "mosaic": "00\n00",
+    },
+    "2row x 2col (1:1)": {
+        "kwargs":
+        "gridspec_kw={\"height_ratios\":[1,1]}",
+        "mosaic": "01\n23",
+    },
+}
+
+
+def _idx2figname(idx):
+    return f"Figure {idx}"
+
+
+def _figname2idx(fname):
+    return int(fname.split(" ")[-1])
+
+
+def _update_ax(fig, ax, single_simulation, plot_type, plot_config):
+    """Refresh plots with simulation_data.
+
+    Parameters
+    ----------
+    fig : Figure
+        A matplotlib.figure.Figure object.
+    ax : Axes
+        matplotlib.axes.Axes
+    single_simulation : dict
+        A single simulation
+    plot_type : str
+        Type of subplots
+    plot_config : dict
+        A dict specifies the preprocessing and style of plots.
+    """
+    # Make sure that visualization does not change the original data
+    dpls_copied = copy.deepcopy(single_simulation['dpls'])
+    net_copied = copy.deepcopy(single_simulation['net'])
+
+    for dpl in dpls_copied:
+        dpl.smooth(plot_config['dipole_smooth']).scale(
+            plot_config['dipole_scaling'])
+
+    if net_copied is None:
+        print("No network data")
+        return
+
+    if plot_type == 'spikes':
+        if net_copied.cell_response:
+            net_copied.cell_response.plot_spikes_raster(ax=ax, show=False)
+
+    elif plot_type == 'input histogram':
+        # BUG: got error here, need a better way to handle exception
+        if net_copied.cell_response:
```

Comment:
Is the exception because `cell_response` is empty? Think this should only be a problem if you run a simulation with no drives

Reply:
I guess so. I'll remove that comment